File: trees/112.js
class Solution {
    inorder(root, sum, targetSum) {
        if (root === null) {
            return false;
        }
        sum += root.val;
        if (root.left === null && root.right === null) {
            return sum === targetSum;
        }
        const leftSide = this.inorder(root.left, sum, targetSum);
        const rightSide = this.inorder(root.right, sum, targetSum);
        return leftSide || rightSide;
    }

    hasPathSum(root, targetSum) {
        let sum = 0;
        const result = this.inorder(root, sum, targetSum);
        return result;
    }
}

File: trees/112.py
class Solution:
    def inorder(self, root, sum, targetSum):
        if root is None:
            return False
        sum += root.val
        if root.left is None and root.right is None:
            return sum == targetSum
        leftSide = self.inorder(root.left, sum, targetSum)
        rightSide = self.inorder(root.right, sum, targetSum)
        return leftSide or rightSide

    def hasPathSum(self, root, targetSum):
        sum = 0
        result = self.inorder(root, sum, targetSum)
        return result